ip_proofs/src/gipa.rs

```diff
@@ -0,0 +1,556 @@
+        if IP::inner_product(values.0, values.1)? != values.2.clone() {
+            return Err(Box::new(InnerProductArgumentError::InnerProductInvalid));
+        }
+        if values.0.len().count_ones() != 1 {
+            // Power of 2 length
+            return Err(Box::new(InnerProductArgumentError::MessageLengthInvalid(
+                values.0.len(),
+                values.1.len(),
+            )));
+        }
+        if !(LMC::verify(ck.0, values.0, com.0)?
+            && RMC::verify(ck.1, values.1, com.1)?
+            && IPC::verify(&vec![ck.2.clone()], &vec![values.2.clone()], com.2)?)
+        {
+            return Err(Box::new(InnerProductArgumentError::InnerProductInvalid));
+        }
+
+        let (proof, _) =
+            Self::prove_with_aux((values.0, values.1), (ck.0, ck.1, &vec![ck.2.clone()]))?;
+        Ok(proof)
+    }
+
+    pub fn verify(
+        ck: (&[LMC::Key], &[RMC::Key], &IPC::Key),
+        com: (&LMC::Output, &RMC::Output, &IPC::Output),
+        proof: &GIPAProof<IP, LMC, RMC, IPC, D>,
+    ) -> Result<bool, Error> {
+        if ck.0.len().count_ones() != 1 || ck.0.len() != ck.1.len() {
+            // Power of 2 length
+            return Err(Box::new(InnerProductArgumentError::MessageLengthInvalid(
+                ck.0.len(),
+                ck.1.len(),
+            )));
+        }
+        let mut clone = Clone::clone(proof);
+        Self::recursive_verify(
+            (ck.0, ck.1, &vec![ck.2.clone()]),
+            com,
+            &mut clone,
+            &Default::default(),
+        )
+    }
+
+    pub fn prove_with_aux(
+        values: (&[IP::LeftMessage], &[IP::RightMessage]),
+        ck: (&[LMC::Key], &[RMC::Key], &[IPC::Key]),
+    ) -> Result<
+        (
+            GIPAProof<IP, LMC, RMC, IPC, D>,
+            GIPAAux<IP, LMC, RMC, IPC, D>,
+        ),
+        Error,
+    > {
+        let (r_commitment_steps, r_base, r_transcript, ck_base) =
+            Self::recursive_prove(values, ck, &Default::default())?;
+
+        Ok((
+            GIPAProof {
+                r_commitment_steps,
+                r_base,
+                _gipa: PhantomData,
+            },
+            GIPAAux {
+                r_transcript,
+                ck_base,
+                _gipa: PhantomData,
+            },
+        ))
+    }
+
+    // Returns vector of recursive commitments and transcripts in reverse order
+    fn recursive_prove(
+        values: (&[IP::LeftMessage], &[IP::RightMessage]),
+        ck: (&[LMC::Key], &[RMC::Key], &[IPC::Key]),
+        transcript: &LMC::Scalar,
+    ) -> Result<
+        (
+            Vec<(
+                (LMC::Output, RMC::Output, IPC::Output),
+                (LMC::Output, RMC::Output, IPC::Output),
+            )>,
+            (LMC::Message, RMC::Message),
+            Vec<LMC::Scalar>,
+            (LMC::Key, RMC::Key),
+        ),
+        Error,
+    > {
+        let (m_a, m_b) = values;
+        let (ck_a, ck_b, ck_t) = ck;
+        match m_a.len() {
```

Comment:
maybe this is can be replaced with a 

```rust
if m_a.len() == 1 { something } else { something else }
```

Reply:
:+1: 

Fixed when unrolling recursion.